python/tests/utils/target_env_var.py

```diff
@@ -0,0 +1,91 @@
+# ============================================================================ #
+# Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                   #
+# All rights reserved.                                                         #
+#                                                                              #
+# This source code and the accompanying materials are made available under     #
+# the terms of the Apache License 2.0 which accompanies this distribution.     #
+# ============================================================================ #
+
+# RUN: PYTHONPATH=../../ pytest -rP  %s
+
+import os
+os.environ["CUDAQ_DEFAULT_SIMULATOR"] = "density-matrix-cpu"
+
+import pytest
+
+import cudaq
+
+def test_default_target():
+    """Tests the default target set by environment variable"""
+
+    assert ("density-matrix-cpu" == cudaq.get_target().name)
+
+    kernel = cudaq.make_kernel()
+    qubits = kernel.qalloc(2)
+    kernel.h(qubits[0])
+    kernel.cx(qubits[0], qubits[1])
+    kernel.mz(qubits)
+
+    result = cudaq.sample(kernel)
+    result.dump()
+    assert '00' in result
+    assert '11' in result
+
+def test_env_var_with_emulate():
+    """Tests the target when emulating a hardware backend"""
+
+    assert ("density-matrix-cpu" == cudaq.get_target().name)
+    cudaq.set_target("quantinuum", emulate=True)
+    assert ("quantinuum" == cudaq.get_target().name)
+    
+    kernel = cudaq.make_kernel()
+    qubits = kernel.qalloc(2)
+    kernel.h(qubits[0])
+    kernel.cx(qubits[0], qubits[1])
+    kernel.mz(qubits)
+
+    result = cudaq.sample(kernel)
+    result.dump()
+    assert '00' in result
+    assert '11' in result
+
+def test_target_override():
+    """Tests the target set by environment variable is overridden by user setting"""
+
+    cudaq.set_target("qpp-cpu")
+    assert("qpp-cpu" == cudaq.get_target().name)
+
+    kernel = cudaq.make_kernel()
+    qubits = kernel.qalloc(2)
+    kernel.h(qubits[0])
+    kernel.cx(qubits[0], qubits[1])
+    kernel.mz(qubits)
+
+    result = cudaq.sample(kernel)
+    result.dump()
+    assert '00' in result
+    assert '11' in result
+
+def test_ignore_invalid_target():
+    """Tests that if the environment variable is not a valid target, it is ignored"""
+
+    os.environ["CUDAQ_DEFAULT_SIMULATOR"] = "foo"
+    assert("foo" != cudaq.get_target().name)
```

Comment:
Hm, is this the expected behavior:

kernel1 = cudaq.make_kernel()
...
os.environ["CUDAQ_DEFAULT_SIMULATOR"] = "nvidia"
cudaq.sample(kernel) # runs on nvidia backend?
os.environ["CUDAQ_DEFAULT_SIMULATOR"] = "tensornet"
cudaq.sample(kernel) # runs on tensornet backend?

If so, it would be good to add this test. As written, the target set from a previous tests could still be active here. However, my expectation would be that since I now have an invalid environment variable, 
either the CPU or GPU statevector simulator should be the target here. Would it make sense to call reset before and after each test to encapsulate the tests better?

Reply:
Also, that's not expected behavior since the environment variable isn't checked for setting current target, only for default target. I will update the tests to showcase the behavior.

Reply:
Well, in my snippet above, there is no explicit target defined, so both invocations of sample run on the default simulator. I think what you are saying is that the environment variable *at the time when cudaq is imported* (correct me if that's wrong) is relevant, not the value of the environment variable *at the time when the simulator is invoked*. I would clarify that in the docs.

Reply:
Yes, you are correct. I will add it to the documentation.